src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs

```diff
@@ -312,6 +353,32 @@ private static void Unscale(ref uint low, ref ulong high64, ref int scale)
                     scale--;
             }
 
+            /// <summary>
+            /// Do partial divide, yielding 64-bit result and 64-bit remainder.
+            /// Divisor must be larger than upper 64 bits of dividend.
+            /// </summary>
+            /// <param name="bufNum">128-bit dividend as array of uints, least-sig first</param>
+            /// <param name="den">64-bit divisor</param>
+            /// <returns>Returns quotient. Remainder overwrites lower 64-bits of dividend.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static unsafe ulong Div128By64(Buf16* bufNum, ulong den)
```

Comment:
Same general comment, but this is notably a case where we'd ideally just use `UInt128.operator /` and have the JIT recognize the general pattern instead.

Doing general purpose optimizations is almost always a better option than doing one offs. Ideally the JIT recognizes cases like `ulong / uint` or `uint128 / ulong` and does the optimal thing on platforms that support it.

Reply:
In general that would be god, but in this specific case there we know several things that the compiler cannot know and optimize for.

* denominator has already been shifted left so it has highest bit set
   * The metod will probably be called several times in a row so it makes sense to do bit shifting etc before it is called
* den > bufNum->High64

However since both types lives in CoreLib, several helpers here could probably be made internal and useable by both UInt128 and Decimal division code. For example I think both this metod and maybe division by 32bit could be usefull from within UInt128

Reply:
I opened #99747 with a similar fix for UInt128.

I did not make this internal and call it from UInt128 at the moment, but if the code works as fast without the divisor having the highest bit set then it might make sense to do so.